test/test_nn.py

```diff
@@ -11351,6 +11390,116 @@ def test_maxpool3d_non_square_backward(self, device):
+                torch.manual_seed(3)
+                bias = None if has_bias else torch.randn((oc), device=input_device, dtype=dtype)
+
+                thnn_output = thnn_conv_group(input, weight, k, bias, s, p, d, g)
+                if device == 'cpu':
+                    output = torch.conv2d(input, weight, bias, s, p, d, g)
+                elif device == 'mkldnn':
+                    output = torch.mkldnn_convolution(input, weight, bias, p, s, d, g)
+                elif device == 'cuda':
+                    output = torch.conv2d(input, weight, bias, s, p, d, g)
+                elif device == 'cudnn':
+                    output = torch.cudnn_convolution(input, weight, bias, p, s, d, g, True, True)
+
+
+                if device == 'cuda' or device == 'cudnn':
+                    try:
+                        self.assertEqual(
+                            output, thnn_output,
+                            message='device:{}, dtype:{}, input size:{}, group:{}, '
+                                    'batchsize:{}, input channel:{}, output channel:{}, '
+                                    'bias:{}, padding:{}, dilation:{}, stride:{}, '
+                                    'kernel:{}'.format(device, dtype, isize, g, bs, ic, 
+                                                       oc, has_bias, p, d, s, k)
+                        )
+                    except AssertionError as e:
+                        print(e)
+                else:
+                    self.assertEqual(
+                        output, thnn_output,
+                        message='device:{}, dtype:{}, input size:{}, group:{}, '
+                                'batchsize:{}, input channel:{}, output channel:{}, '
+                                'bias:{}, padding:{}, dilation:{}, stride:{}, '
+                                'kernel:{}'.format(device, dtype, isize, g, bs, ic, 
+                                                   oc, has_bias, p, d, s, k)
+                    )
+
+        count = 0
+        for case in [test_case1, test_case2, test_case3]:
+            case_combinations = list(product(*map(lambda x: x[1], case)))
+
+            for combination in case_combinations:
+                group = combination[0]
+                bs = combination[1]
+                ic = combination[2]
+                oc = combination[3]
+                p = combination[4]
+                d = combination[5]
+                s = combination[6]
+                k = combination[7]
+                osize = combination[8]
+                bias = combination[9]
+
+                count += 1
+                isize = []
+                for i in range(len(osize)):
+                    isize.append(s[i] * (osize[i] - 1) + 1 + d[i] * (k[i] - 1) - 2 * p[i])
+
+                if isize[0] > 0 and isize[1] > 0:
+                    _test_conv2d_compare_with_thnn(group, bs, isize, ic, oc, bias, p, d, s, k, 'cpu')
+                    if torch.backends.mkldnn.is_available():
```

Comment:
You can clarify and simplify your code here by making this class generic. See how TestNNDeviceType is handled. This will give you access to decorators like @onlyOnCPUAndCUDA, @@skipCUDAIfRocm, @dtypes, and @@dtypesIfCUDA. See test_blas_alpha_beta_empty in test_torch.py for an example of what a test looks like. This wiki article has more information: https://github.com/pytorch/pytorch/wiki/Writing-tests-that-run-on-all-available-device-types.